Threaded blocksparse

mtfishman opened the pull request:
This adds multithreading to block sparse contractions using Julia's native multithreading (`Threads.@spawn`). The basic interface is as follows. Start Julia with more than one thread with:
```julia
julia -t 4
```
or
```julia
JULIA_NUM_THREADS=4 julia
```
Then, enable block sparse multithreading with `NDTensors.enable_threaded_blocksparse()` (you should probably also turn off BLAS and Strided multithreading, which we could automate within NDTensors):
```julia
using ITensors
using LinearAlgebra

BLAS.set_num_threads(1)
ITensors.Strided.disable_threads()

@show Threads.nthreads()

N = 4
d = 1000
i = Index([QN(0, N) => d for n in 1:N], "i")

# Matrix multiplication
A = randomITensor(i'', dag(i'))
B = randomITensor(i', dag(i))

ITensors.disable_threaded_blocksparse()
@show ITensors.using_threaded_blocksparse()
C1 = @time A * B

ITensors.enable_threaded_blocksparse()
@show ITensors.using_threaded_blocksparse()
C2 = @time A * B
ITensors.disable_threaded_blocksparse()

@show C1 ≈ C2
```
returns:
```julia
Threads.nthreads() = 4
ITensors.using_threaded_blocksparse() = false
  2.204602 seconds (508 allocations: 122.121 MiB)
ITensors.using_threaded_blocksparse() = true
  0.688744 seconds (1.76 k allocations: 122.205 MiB)
C1 ≈ C2 = true
```

mtfishman commented:
Results for 2D Hubbard with momentum conservation:

With 1 thread:
```
nnz(H[end ÷ 2]) = 120
After sweep 1 energy=-5.680481085305 maxlinkdim=64 time=1.074
After sweep 2 energy=-11.588145697827 maxlinkdim=200 time=8.511
After sweep 3 energy=-13.418083112850 maxlinkdim=400 time=16.430
After sweep 4 energy=-13.584293087750 maxlinkdim=800 time=12.010
After sweep 5 energy=-13.659583110597 maxlinkdim=2000 time=18.421
After sweep 6 energy=-13.700596011172 maxlinkdim=2999 time=29.308
After sweep 7 energy=-13.712935518643 maxlinkdim=3000 time=34.207
After sweep 8 energy=-13.715789709726 maxlinkdim=3000 time=36.194
After sweep 9 energy=-13.716399114988 maxlinkdim=3000 time=36.783
After sweep 10 energy=-13.716546234503 maxlinkdim=2999 time=35.400
228.430532 seconds (1.34 G allocations: 377.864 GiB, 9.28% gc time)
```

With 5 threads:
```
nnz(H[end ÷ 2]) = 120
After sweep 1 energy=-5.680481085305 maxlinkdim=64 time=0.740
After sweep 2 energy=-11.588145697827 maxlinkdim=200 time=6.592
After sweep 3 energy=-13.418083112850 maxlinkdim=400 time=12.815
After sweep 4 energy=-13.584293087750 maxlinkdim=800 time=8.853
After sweep 5 energy=-13.659583110597 maxlinkdim=2000 time=12.636
After sweep 6 energy=-13.700596011172 maxlinkdim=2999 time=18.046
After sweep 7 energy=-13.712935518643 maxlinkdim=3000 time=19.852
After sweep 8 energy=-13.715789709726 maxlinkdim=3000 time=20.499
After sweep 9 energy=-13.716399114988 maxlinkdim=3000 time=20.423
After sweep 10 energy=-13.716546234503 maxlinkdim=2999 time=20.106
140.651422 seconds (1.37 G allocations: 378.625 GiB, 16.96% gc time)
```

The same calculation with C++ ITensor with 1 thread:
```
    vN Entropy at center bond b=9 = 3.561531540785
    Eigs at center bond b=9: 0.1118 0.1117 0.0949 0.0949 0.0544 0.0543 0.0543 0.0543 0.0246 0.0246
    Largest link dim during sweep 10/10 was 3000
    Largest truncation error: 2.27609e-07
    Energy after sweep 10/10 is -13.716559596412
    Sweep 10/10 CPU time = 39.08s (Wall time = 43.97s)
```
and C++ ITensor with 5 threads:
```
    vN Entropy at center bond b=9 = 3.561531540785
    Eigs at center bond b=9: 0.1118 0.1117 0.0949 0.0949 0.0544 0.0543 0.0543 0.0543 0.0246 0.0246
    Largest link dim during sweep 10/10 was 3000
    Largest truncation error: 2.27609e-07
    Energy after sweep 10/10 is -13.716559596412
    Sweep 10/10 CPU time = 1m, 42.4s (Wall time = 22.13s)
```

mtfishman commented:
The timings are updated with a new version of NDTensors. You can see that the single threaded code is quite fast at this point, and the speedup from multithreading is not as pronounced compared to the C++ version. Getting a better speedup may require some more advanced multithreading techniques (for example handling the case of repeated contractions to the same block better). Speedups may be better for larger system sizes with more blocks.

Note that this multithreading doesn't seem to compose well with BLAS multithreading or Strided multithreading. In principle, it should compose with Strided multithreading, since they are both using Julia's native `@spawn` framework, but that needs to be investigated. A question that arises is if we should somehow automatically turn of BLAS and Strided multithreading when NDTensors multithreading is being used. In principle it would be nice, but in practice it becomes a question of where to turn it off (i.e. globally or locally only where our threading is being used).

Anyway, I think this is good to go for now, and we can iterate on it as needed.

emstoudenmire commented:
It is already giving a worthwhile speedup, so then I think it's good to go and we could always iterate on it more. The main thing I wanted to see is that the defaults make sense (e.g. you still have to opt in to this even if running Julia in a multithreaded environment, which is what we want for now). So then in the docs we can guide people through the best practices like turning off BLAS multithreading etc.

One minor question is: I see timings for 5 threads above. What's the speedup over single-threaded (including single-threaded BLAS and Strided) when just using two threads, say?

mtfishman commented:
> It is already giving quite a worthwhile speedup, so then I think it's good to go and we could always iterate on it more. The main thing I wanted to see is that the defaults make sense (e.g. you still have to opt in to this even if running Julia in a multithreaded environment, which is what we want for now). So then in the docs we can guide people through the best practices like turning off BLAS multithreading etc.

Sounds good. Indeed, block sparse threading is off by default even if Julia is started with multiple threads.

> 
> One minor question is: I see timings for 5 threads above. What's the speedup over single-threaded (including single-threaded BLAS and Strided) when just using two threads, say?

Good question, I'll check.

mtfishman commented:
I've updated the second comment with results for 2 threads. I also rewrote the loop with partitioning, which improved the results a bit for 5 threads.

For future reference, here is a comparison between different ways to multithread a for loop with native Julia multithreading:
```julia
using BenchmarkTools

function f1!(y, x)
  for n in eachindex(x)
    @inbounds y[n] = sin(x[n])
  end
  return y
end

function f2!(y, x)
  Threads.@threads for n in eachindex(x)
    @inbounds y[n] = sin(x[n])
  end
  return y
end

function f3!(y, x)
  @sync for n in eachindex(x)
    Threads.@spawn begin
      @inbounds y[n] = sin(x[n])
    end
  end
  return y
end

function f4!(y, x)
  @sync for n_partition in Iterators.partition(eachindex(x), max(1, length(eachindex(x)) ÷ Threads.nthreads()))
    Threads.@spawn for n in n_partition
      @inbounds y[n] = sin(x[n])
    end
  end
  return y
end

N = 20_000
x = randn(N)
y = similar(x)
@btime f1!($y, $x)
@btime f2!($y, $x)
@btime f3!($y, $x)
@btime f4!($y, $x)
```
where with 4 threads I get:
```
  187.558 μs (0 allocations: 0 bytes)
  47.332 μs (21 allocations: 2.98 KiB)
  8.680 ms (140025 allocations: 15.45 MiB)
  51.871 μs (36 allocations: 3.56 KiB)
```
You can see that `f2!` and `f4!` get nearly perfect speedup, while `f3!` gets no speedup (and is in fact worse than not threading). `f2!` uses the higher level `Threads.@threads` macro that automatically splits up the for loop into chunks, and those chunks are then sent to different threads. In `f3!`, each time the inner part of the loop is being called, it is spawning a new piece of work for Julia to allocate to a thread. However, right now, this has quite a bit of overhead:
```julia
julia> @btime fetch(Threads.@spawn sin(1))
  3.106 μs (6 allocations: 752 bytes)
0.8414709848078965

julia> @btime sin(1)
  1.180 ns (0 allocations: 0 bytes)
0.8414709848078965
```
so this is much slower. To fix this, in `f4!`, first we split up the outer for loop into partitions of work, and those partitions are spawned as chunks of work for Julia to distribute to threads, so `@spawn` is only called once per thread and the overhead is minimized. Notice that `@sync` is necessary around the for loop to indicate to Julia that we want to wait for all of the work to be finished before moving on from the for loop: https://stackoverflow.com/questions/61905127/what-is-the-difference-between-threads-spawn-and-threads-threads

Essentially, the latest change was rewriting the loop from being written like `f3!` to being written like `f4!` (though there was already some level of partitioning being done, so the difference is not so drastic as in this toy example).

emstoudenmire commented:
Thanks Matt. These comments and findings will be useful to refer back to as we do multithreading in other places. It will be interesting to see how composable the multithreading is if, say, we multithread over multiple MPS in a sum, while simultaneously multithreading over QNITensor blocks.
